src/metabase/driver.clj

```diff
@@ -292,11 +292,37 @@
   [date-format-str]
   (.withOffsetParsed ^DateTimeFormatter (tformat/formatter date-format-str)))
 
+(defn- exception-as-return-value
+  "Invoke `f`, return an exception rather than throwing it"
+  [f]
+  (try
+    (f)
+    (catch Exception e
+      e)))
+
+(defn- first-successful-parse
+  "Attempt to parse `time-str` with each of `date-formatters`, returning the first successful parse. If there are no
+  successful parses return the most recent Exception returned."
+  [date-formatters time-str]
+  (loop [[df & more-df] date-formatters]
+    (let [result (exception-as-return-value #(tformat/parse df time-str))]
+      (cond
+        (and (instance? Exception result)
+             (seq more-df))
+        (recur more-df)
+
+        (and (instance? Exception result)
+             (empty? more-df))
+        (throw result)
+
+        :else
+        result))))
```

Comment:
Would it be more readable if you just wrote this whole thing like 

```clojure
(defn- first-successful-parse
  [date-formatters time-str]
  (let [last-exception (atom nil)]
    (or (some #(try (% time-str)
                    (catch Throwable e
                      (reset! last-exception e)))
              date-formatters)
        @last-exception)))
```

Or maybe even like (possibly overly clever)

```clojure
(defn- first-successful-parse
  [date-formatters time-str]
  (or (some #(u/ignore-exceptions (% time-str)) date-formatters)
      (try
        (doseq [formatter date-formatters] (formatter time-str))
        (catch Throwable e
          e))))
```

Reply:
Maybe something like
``` clojure
(defn- attempt-date-parse
  "Invoke `f`, return an exception rather than throwing it"
  [date-formatter time-str]
  (try
    (tformat/parse formatter time-str)
    (catch Exception e
      e)))

(defn- first-successful-parse
  "Attempt to parse `time-str` with each of `date-formatters`, returning the first successful parse. If there are no
  successful parses throws the exception that the last formatter threw."
  [date-formatters time-str]
  (let [applied-formatters (map #(attempt-date-parse % time-str) date-formatters)
        exception?         #(instance? Exception %)]
    (or (some (complement exception?) applied-formatters)
        (throw (some exception? (reverse applied-formatters))))))
```

Reply:
`some` returns the value returned by its `f` argument rather than the `x` that caused `(f x)` to be truthy so this function would just end up returning `true` if it succeeded. Also if a date parser returned `nil` it would be considering that to be valid

Reply:
Yikes, let's try this, should have used a remove/filter combo
``` clojure
    (or (remove exception? applied-formatters)
        (throw (filter exception? (reverse applied-formatters))))
```

Reply:
Maybe you could do something like

```clojure
(defn- first-successful-parse
  "Attempt to parse `time-str` with each of `date-formatters`, returning the first successful parse. If there are no
  successful parses throws the exception that the last formatter threw."
  [date-formatters time-str]
  (let [results (filter some? (for [formatter date-formatters]
                                (try (formatter time-str)
                                     (catch Throwable e
                                       e))))]
    (or (some #(when-not (instance? Exception %) %) results)
        (doseq [result (reverse results)
                :when (instance? Exception result)]
          (throw result)))))
```

Reply:
👍 on the `doseq`, I'll switch it to that

Reply:
@senior ok I got it.

```clojure
(defn- first-successful-parse
  "Attempt to parse `time-str` with each of `date-formatters`, returning the first successful parse. If there are no
  successful parses throws the exception that the last formatter threw."
  [date-formatters time-str]
  (let [{exceptions true, results false} (group-by
                                          (partial instance? Exception)
                                          (for [formatter date-formatters]
                                            (try (formatter time-str)
                                                 (catch Throwable e
                                                   e))))]
    (or (some identity results)
        (when-let [[last-exception] (seq (reverse exceptions))]
          (throw last-exception)))))
```